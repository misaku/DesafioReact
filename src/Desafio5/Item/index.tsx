import React from "react";
import { useItemController } from "./useItemController";

interface ItemProps {
  valor: number;
  parOuImpar(num: number): string;
}
export const Item: React.FC<ItemProps> = ({ valor, parOuImpar }) => {
  const { renderCounter, count, addCount } = useItemController();
  return (
    <li>
      <button onClick={addCount}>
        adiciona + 1 na contagem do item {valor}
      </button>
      <p>
        contagem: {count} Ã© {parOuImpar(count)}
      </p>
      <p>renderizou: {renderCounter}</p>
    </li>
  );
};
