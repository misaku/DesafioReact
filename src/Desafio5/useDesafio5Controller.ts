import { useState } from "react";

export const useDesafio5Controller = () => {
  const [list, setList] = useState<number[]>([]);
  const addItem = () => setList([...list, list.length]);
  const isPar = (valor: number) => valor % 2 === 0;
  const numero = list.length + 1;
  const parOuImpar = (valor: number) => (isPar(valor) ? "Par" : "Impar");

  return { list, addItem, numero, parOuImpar };
};
