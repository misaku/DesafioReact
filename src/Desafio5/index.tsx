import React from "react";
import { Item } from "./Item";
import { useDesafio5Controller } from "./useDesafio5Controller";

/* TODO
  O componente abaixo está com problemas. Resolva o problema e melhore a performance.

  Obs. considere o componente filho também para resolver esse desafio

  Comentários sobre a sua solução:

*/

export function Desafio5() {
  const { list, addItem, numero, parOuImpar } = useDesafio5Controller();
  return (
    <>
      <button onClick={addItem}  data-testid="btn1">adiciona + 1 item na lista</button>
      <span data-testid="size">{list.length}</span>
      <ul>
        {list.map((item) => (
          <Item valor={item} parOuImpar={parOuImpar} />
        ))}
      </ul>
      <p>
        {numero} é {parOuImpar(numero)}
      </p>
    </>
  );
}
