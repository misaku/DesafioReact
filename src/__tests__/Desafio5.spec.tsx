import {Desafio5} from '../Desafio5';
import {waitFor, render, getNodeText} from '@testing-library/react'
import user from '@testing-library/user-event'
describe('Desafio5',()=>{
  it('Deve ser Renderizado com os valores corretos 1',async ()=>{
    const  component = render(<Desafio5 />)
    const btn1 =await component.getByTestId('btn1')
    const size =await component.getByTestId('size')
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await waitFor(()=>expect(getNodeText(size)).toEqual('8'))
    expect(component).toMatchSnapshot()
  })
  it('Deve ser Renderizado com os valores correto 2',async ()=>{
    const  component = render(<Desafio5 />)
    const btn1 =await component.getByTestId('btn1')
    const size =await component.getByTestId('size')
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await waitFor(()=>expect(getNodeText(size)).toEqual('10'))
    expect(component).toMatchSnapshot()
  })
})
