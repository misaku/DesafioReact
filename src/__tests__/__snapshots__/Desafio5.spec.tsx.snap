// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Desafio5 Deve ser Renderizado com os valores correto 2 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        data-testid="btn1"
      >
        adiciona + 1 item na lista
      </button>
      <span
        data-testid="size"
      >
        10
      </span>
      <ul>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            0
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            1
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            2
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            3
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            4
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            5
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            6
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            7
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            8
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            9
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
      </ul>
      <p>
        11
         é  
        Impar
      </p>
    </div>
  </body>,
  "container": <div>
    <button
      data-testid="btn1"
    >
      adiciona + 1 item na lista
    </button>
    <span
      data-testid="size"
    >
      10
    </span>
    <ul>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          0
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          1
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          2
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          3
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          4
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          5
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          6
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          7
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          8
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          9
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
    </ul>
    <p>
      11
       é  
      Impar
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Desafio5 Deve ser Renderizado com os valores corretos 1 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        data-testid="btn1"
      >
        adiciona + 1 item na lista
      </button>
      <span
        data-testid="size"
      >
        8
      </span>
      <ul>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            0
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            1
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            2
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            3
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            4
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            5
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            6
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
        <li>
          <button>
            adiciona + 1 na contagem do item 
            7
          </button>
          <p>
            contagem: 
            0
             é 
            Par
          </p>
          <p>
            renderizou: 
            1
          </p>
        </li>
      </ul>
      <p>
        9
         é  
        Impar
      </p>
    </div>
  </body>,
  "container": <div>
    <button
      data-testid="btn1"
    >
      adiciona + 1 item na lista
    </button>
    <span
      data-testid="size"
    >
      8
    </span>
    <ul>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          0
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          1
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          2
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          3
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          4
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          5
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          6
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
      <li>
        <button>
          adiciona + 1 na contagem do item 
          7
        </button>
        <p>
          contagem: 
          0
           é 
          Par
        </p>
        <p>
          renderizou: 
          1
        </p>
      </li>
    </ul>
    <p>
      9
       é  
      Impar
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
